File: train/input.py
#!/usr/bin/env python3
"""
train/input.py — Position Encoder  
Converts chess positions to 8×8×19 spatial tensors for the ResNet.

Planes:
  [0..11]  - 12 piece planes (White P/N/B/R/Q/K, Black P/N/B/R/Q/K)
  [12..15] - 4 castling rights (WK, WQ, BK, BQ)
  [16]     - en passant file (plane of 1.0 on ep-file column)
  [17]     - side to move (all 1.0 = white, all 0.0 = black)
  [18]     - ones bias plane (always 1.0)

Move index: from_sq * 73 + to_idx
  to_idx: 0-63 = target square, 64-67 = promo (N/B/R/Q)
  Max index: 63 * 73 + 72 = 4671  →  POLICY_SIZE = 4672

Square convention (matching TypeScript engine):
  sq 0 = a8 (top-left), sq 63 = h1 (bottom-right)
  row = sq // 8, col = sq % 8
  chess.py uses sq 0 = a1 (bottom-left), so we flip rank:
    our_sq = (7 - chess.square_rank(chess_sq)) * 8 + chess.square_file(chess_sq)
"""

import chess
import numpy as np
import torch
from typing import Optional

# ── Constants ─────────────────────────────────────────────────────────────────

N_PLANES    = 19
POLICY_SIZE = 4672
BOARD_SIZE  = 8

# chess.py piece type → plane index (0-5 for white, 6-11 for black)
_PIECE_PLANE = {
    chess.PAWN: 0, chess.KNIGHT: 1, chess.BISHOP: 2,
    chess.ROOK: 3, chess.QUEEN:  4, chess.KING:   5,
}

_PROMO_OFFSET = {
    chess.KNIGHT: 64,
    chess.BISHOP: 65,
    chess.ROOK:   66,
    chess.QUEEN:  67,  # queen promotion = normal to target sq (no special offset)
}


# ── Square conversion ─────────────────────────────────────────────────────────

def chess_sq_to_our_sq(chess_sq: int) -> int:
    """Convert chess.py square (a1=0) to our convention (a8=0)."""
    return (7 - chess.square_rank(chess_sq)) * 8 + chess.square_file(chess_sq)


def our_sq_to_chess_sq(our_sq: int) -> int:
    """Convert our square (a8=0) to chess.py square (a1=0)."""
    row = our_sq // 8
    col = our_sq % 8
    return chess.square(col, 7 - row)


# ── Board encoding ────────────────────────────────────────────────────────────

def encode_board(board: chess.Board) -> np.ndarray:
    """
    Encode a chess.Board as float32[19, 8, 8] tensor.
    Always encodes from the perspective of the SIDE TO MOVE (board.turn).
    
    Returns numpy array [19, 8, 8].
    """
    planes = np.zeros((N_PLANES, BOARD_SIZE, BOARD_SIZE), dtype=np.float32)
    
    for chess_sq in range(64):
        piece = board.piece_at(chess_sq)
        if piece is None:
            continue
        
        plane_idx = _PIECE_PLANE[piece.piece_type]
        if piece.color == chess.BLACK:
            plane_idx += 6
        
        our_sq = chess_sq_to_our_sq(chess_sq)
        row, col = our_sq // 8, our_sq % 8
        planes[plane_idx, row, col] = 1.0
    
    # Castling rights (planes 12-15)
    if board.has_kingside_castling_rights(chess.WHITE):   planes[12] = 1.0
    if board.has_queenside_castling_rights(chess.WHITE):  planes[13] = 1.0
    if board.has_kingside_castling_rights(chess.BLACK):   planes[14] = 1.0
    if board.has_queenside_castling_rights(chess.BLACK):  planes[15] = 1.0
    
    # En passant file (plane 16)
    if board.ep_square is not None:
        ep_col = chess.square_file(board.ep_square)
        planes[16, :, ep_col] = 1.0
    
    # Side to move (plane 17): 1.0 = white to move
    if board.turn == chess.WHITE:
        planes[17] = 1.0
    
    # Bias plane (plane 18): always 1.0
    planes[18] = 1.0
    
    return planes


def encode_batch(boards: list[chess.Board]) -> torch.Tensor:
    """Encode multiple boards into a batched tensor [N, 19, 8, 8]."""
    batch = np.stack([encode_board(b) for b in boards], axis=0)
    return torch.from_numpy(batch)


# ── Move encoding ─────────────────────────────────────────────────────────────

def move_to_index(move: chess.Move) -> int:
    """
    Convert a chess.Move to a policy index in [0, 4671].
    Queen promotion uses the normal to-square (no promo offset).
    Under-promotions (N/B/R) use offset 64-66.
    """
    from_sq = chess_sq_to_our_sq(move.from_square)
    
    if move.promotion and move.promotion != chess.QUEEN:
        return from_sq * 73 + _PROMO_OFFSET[move.promotion]
    
    to_sq = chess_sq_to_our_sq(move.to_square)
    return from_sq * 73 + to_sq


def index_to_move_hint(idx: int) -> tuple[int, int, Optional[int]]:
    """
    Convert policy index to (from_our_sq, to_our_sq, promo_piece).
    promo_piece is a chess.* piece type constant or None.
    """
    from_sq = idx // 73
    rest    = idx % 73
    
    if rest == 64: return from_sq, -1, chess.KNIGHT
    if rest == 65: return from_sq, -1, chess.BISHOP
    if rest == 66: return from_sq, -1, chess.ROOK
    if rest == 67: return from_sq, -1, chess.QUEEN
    return from_sq, rest, None


def get_legal_move_indices(board: chess.Board) -> tuple[list[chess.Move], list[int]]:
    """Return (legal_moves, corresponding_policy_indices) for a board."""
    moves   = list(board.legal_moves)
    indices = [move_to_index(m) for m in moves]
    return moves, indices


# ── Game outcome ──────────────────────────────────────────────────────────────

def game_result_to_value(result: str, turn: chess.Color) -> float:
    """
    Convert PGN result string to value from the perspective of `turn`.
    '1-0' = white wins, '0-1' = black wins, '1/2-1/2' = draw
    """
    if result == '1-0':     outcome =  1.0  # white wins
    elif result == '0-1':   outcome = -1.0  # black wins
    else:                   outcome =  0.0  # draw
    
    return outcome if turn == chess.WHITE else -outcome


if __name__ == '__main__':
    # Sanity check
    board = chess.Board()
    planes = encode_board(board)
    assert planes.shape == (19, 8, 8), f"Shape mismatch: {planes.shape}"
    
    # Verify white pieces on starting position
    # White pawns on row 6 (rank 2), black pawns on row 1 (rank 7)
    assert planes[0, 6, :].sum() == 8, "Expected 8 white pawns on row 6"
    assert planes[6, 1, :].sum() == 8, "Expected 8 black pawns on row 1"
    
    # Move encoding
    e2e4 = chess.Move.from_uci('e2e4')
    idx  = move_to_index(e2e4)
    assert 0 <= idx < POLICY_SIZE, f"Index out of range: {idx}"
    
    # Batch encoding
    batch = encode_batch([board, board])
    assert batch.shape == (2, 19, 8, 8)
    
    print(f"✅ Input encoding OK — planes: {planes.shape}, dtype: {planes.dtype}")
    print(f"   e2e4 index: {idx}  (range: 0-{POLICY_SIZE-1})")

File: src/engine/nn/input.ts
// ── Position → Neural Network Input Encoder ───────────────────────────────────
// Converts a chess position to a Float32Array[19×8×8] spatial tensor.
//
// Planes (19 total):
//   [0..5]   — White pieces: P N B R Q K
//   [6..11]  — Black pieces: P N B R Q K
//   [12]     — White kingside castling (all-1 or all-0 plane)
//   [13]     — White queenside castling
//   [14]     — Black kingside castling
//   [15]     — Black queenside castling
//   [16]     — En passant file (column of ones)
//   [17]     — Side to move (all-1 = white, all-0 = black)
//   [18]     — Bias plane (always all-1)
//
// Square convention: sq 0=a8 (top-left), sq 63=h1 (bottom-right)
// Input tensor shape: [1, 19, 8, 8] — matches Python train/input.py exactly

export const N_PLANES    = 19;
export const BOARD_CELLS = 64;
export const NN_INPUT_SIZE = N_PLANES * BOARD_CELLS; // 1216 (for flat compat)

// Plane indices: WHITE pieces 0-5, BLACK pieces 6-11
// P=0, N=1, B=2, R=3, Q=4, K=5  (×2 for color)
const PIECE_PLANE: Record<number, number> = {
  1: 0, // W_PAWN
  2: 1, // W_KNIGHT
  3: 2, // W_BISHOP
  4: 3, // W_ROOK
  5: 4, // W_QUEEN
  6: 5, // W_KING
  7: 6, // B_PAWN
  8: 7, // B_KNIGHT
  9: 8, // B_BISHOP
 10: 9, // B_ROOK
 11:10, // B_QUEEN
 12:11, // B_KING
};

/**
 * Encode a board position as Float32Array[19×8×8] for ResNet inference.
 * Output layout: plane-major — features[plane * 64 + sq]
 *
 * @param board    Uint8Array[64] piece placement (0=empty, 1-12=pieces)
 * @param turn     0=white, 1=black
 * @param castling CR_* bitmask: bit0=WK, bit1=WQ, bit2=BK, bit3=BQ
 * @param epSq     En passant target square index (-1 = none)
 */
export function encodePosition(
  board: Uint8Array,
  turn: number,
  castling: number,
  epSq: number,
): Float32Array {
  const input = new Float32Array(N_PLANES * BOARD_CELLS);

  // Piece planes 0-11
  for (let sq = 0; sq < 64; sq++) {
    const p = board[sq];
    if (p !== 0) {
      const plane = PIECE_PLANE[p];
      if (plane !== undefined) {
        input[plane * 64 + sq] = 1;
      }
    }
  }

  // Castling planes 12-15 (fill entire 8×8 plane with 1 or 0)
  if (castling & 1) input.fill(1, 12 * 64, 13 * 64); // WK
  if (castling & 2) input.fill(1, 13 * 64, 14 * 64); // WQ
  if (castling & 4) input.fill(1, 14 * 64, 15 * 64); // BK
  if (castling & 8) input.fill(1, 15 * 64, 16 * 64); // BQ

  // En passant file plane 16 (fill entire column)
  if (epSq >= 0) {
    const col = epSq & 7;
    for (let row = 0; row < 8; row++) {
      input[16 * 64 + row * 8 + col] = 1;
    }
  }

  // Side-to-move plane 17 (all-1 = white, all-0 = black)
  if (turn === 0) input.fill(1, 17 * 64, 18 * 64);

  // Bias plane 18 (always all-1)
  input.fill(1, 18 * 64, 19 * 64);

  return input;
}

/**
 * Encode a ChessEngine position into Float32Array[19×8×8].
 */
export function encodeChessEngine(chess: {
  getPosition(): { board: Uint8Array; turn: number; castling: number; epSq: number };
}): Float32Array {
  const pos = chess.getPosition();
  return encodePosition(pos.board, pos.turn, pos.castling, pos.epSq);
}

// ── Move index encoding ────────────────────────────────────────────────────────
// Policy space: from_sq * 73 + to_idx
// from_sq ∈ [0,63], to_idx ∈ [0,72] (0-63=target square, 64-67=promo N/B/R/Q)
// Max index: 63 * 73 + 72 = 4671 → POLICY_SIZE = 4672

export const POLICY_SIZE = 4672;

const PROMO_OFFSET: Record<string, number> = { n: 64, b: 65, r: 66, q: 67 };

function sqFromAlg(alg: string): number {
  const col = alg.charCodeAt(0) - 97;
  const row = 8 - parseInt(alg[1]);
  return row * 8 + col;
}

/**
 * Convert a LAN move string (e.g. "e2e4", "a7a8q") to a policy index (0..4671).
 * Returns -1 if the move format is unrecognised.
 */
export function moveToIndex(lan: string): number {
  if (lan.length < 4) return -1;
  const from  = sqFromAlg(lan.slice(0, 2));
  const to    = sqFromAlg(lan.slice(2, 4));
  const promo = lan[4];
  if (promo && PROMO_OFFSET[promo] !== undefined) {
    return from * 73 + PROMO_OFFSET[promo];
  }
  return from * 73 + to;
}

/**
 * Convert policy index back to a LAN prefix for matching with legal moves.
 */
export function indexToMove(idx: number): { from: number; to: number; promo?: string } {
  const from = Math.floor(idx / 73);
  const rest = idx % 73;
  if (rest >= 64) {
    const promos = ['n', 'b', 'r', 'q'];
    return { from, to: -1, promo: promos[rest - 64] };
  }
  return { from, to: rest };
}
